File: apps/worker/src/relay/balancer.ts
import { GroupMode } from "@wheel/core"

interface GroupItem {
  id: number
  channelId: number
  modelName: string
  priority: number
  weight: number
}

// In-memory round-robin counters (per group)
const rrCounters = new Map<number, number>()

/**
 * Sort and order GroupItems based on the group's load-balancing mode.
 * Returns an ordered array of items to try sequentially.
 */
export function selectChannelOrder(
  mode: GroupMode,
  items: GroupItem[],
  groupId: number,
): GroupItem[] {
  if (items.length === 0) return []

  switch (mode) {
    case GroupMode.RoundRobin:
      return roundRobin(items, groupId)
    case GroupMode.Random:
      return randomOrder(items)
    case GroupMode.Failover:
      return failoverOrder(items)
    case GroupMode.Weighted:
      return weightedOrder(items)
    default:
      return items
  }
}

function roundRobin(items: GroupItem[], groupId: number): GroupItem[] {
  const idx = (rrCounters.get(groupId) ?? 0) % items.length
  rrCounters.set(groupId, idx + 1)
  // Rotate: start from idx, then wrap around
  return [...items.slice(idx), ...items.slice(0, idx)]
}

function randomOrder(items: GroupItem[]): GroupItem[] {
  const shuffled = [...items]
  for (let i = shuffled.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1))
    ;[shuffled[i], shuffled[j]] = [shuffled[j], shuffled[i]]
  }
  return shuffled
}

function failoverOrder(items: GroupItem[]): GroupItem[] {
  // Sort by priority ascending (lower = higher priority)
  return [...items].sort((a, b) => a.priority - b.priority)
}

function weightedOrder(items: GroupItem[]): GroupItem[] {
  // Weighted random: pick one item based on weight, then append the rest
  const totalWeight = items.reduce((sum, it) => sum + it.weight, 0)
  if (totalWeight === 0) return items

  let rand = Math.random() * totalWeight
  let selected = 0
  for (let i = 0; i < items.length; i++) {
    rand -= items[i].weight
    if (rand <= 0) {
      selected = i
      break
    }
  }

  // Put the selected item first, then the rest in order
  const result = [items[selected]]
  for (let i = 0; i < items.length; i++) {
    if (i !== selected) result.push(items[i])
  }
  return result
}

File: apps/worker/internal/relay/balancer.go
package relay

import (
	"math/rand"
	"sort"
	"sync"

	"github.com/kunish/wheel/apps/worker/internal/types"
)

// rrCounters tracks round-robin position per group (goroutine-safe).
var rrCounters sync.Map

// SelectChannelOrder returns group items ordered according to the load-balancing mode.
func SelectChannelOrder(mode types.GroupMode, items []types.GroupItem, groupID int) []types.GroupItem {
	if len(items) == 0 {
		return items
	}

	switch mode {
	case types.GroupModeRoundRobin:
		return roundRobin(items, groupID)
	case types.GroupModeRandom:
		return randomOrder(items)
	case types.GroupModeFailover:
		return failoverOrder(items)
	case types.GroupModeWeighted:
		return weightedOrder(items)
	default:
		return items
	}
}

func roundRobin(items []types.GroupItem, groupID int) []types.GroupItem {
	val, _ := rrCounters.LoadOrStore(groupID, 0)
	idx := val.(int) % len(items)
	rrCounters.Store(groupID, idx+1)

	result := make([]types.GroupItem, 0, len(items))
	result = append(result, items[idx:]...)
	result = append(result, items[:idx]...)
	return result
}

func randomOrder(items []types.GroupItem) []types.GroupItem {
	shuffled := make([]types.GroupItem, len(items))
	copy(shuffled, items)
	// Fisher-Yates shuffle
	for i := len(shuffled) - 1; i > 0; i-- {
		j := rand.Intn(i + 1)
		shuffled[i], shuffled[j] = shuffled[j], shuffled[i]
	}
	return shuffled
}

func failoverOrder(items []types.GroupItem) []types.GroupItem {
	sorted := make([]types.GroupItem, len(items))
	copy(sorted, items)
	sort.Slice(sorted, func(i, j int) bool {
		return sorted[i].Priority < sorted[j].Priority
	})
	return sorted
}

func weightedOrder(items []types.GroupItem) []types.GroupItem {
	totalWeight := 0
	for _, it := range items {
		totalWeight += it.Weight
	}
	if totalWeight == 0 {
		return items
	}

	r := rand.Float64() * float64(totalWeight)
	selected := 0
	for i, it := range items {
		r -= float64(it.Weight)
		if r <= 0 {
			selected = i
			break
		}
	}

	result := make([]types.GroupItem, 0, len(items))
	result = append(result, items[selected])
	for i, it := range items {
		if i != selected {
			result = append(result, it)
		}
	}
	return result
}
